### ISSUE TITLE
Recursive function loses type in 1.1.247

### ISSUE THREAD

**Issue by dbrattli** [microsoft/pyright#3481]
**Describe the bug**

Pyright 1.1.247 looks to make types unknown when calling recursive functions. I cannot see why the return type should become Unknown here, but would appreciate any feedback if I my code is wrong.

**To Reproduce**
Steps to reproduce the behavior. See repro code below. Error is

<img width="1054" alt="Screenshot 2022-05-19 at 08 07 54" src="https://user-images.githubusercontent.com/849479/169222293-493d66e7-a287-4e4b-a93b-99fb01e18e6d.png">

**Expected behavior**

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
from dataclasses import dataclass
from typing import Callable, Generic, Optional, Tuple, TypeVar

Key = TypeVar("Key")
Value = TypeVar("Value")
Result = TypeVar("Result")


@dataclass
class MapTreeLeaf(Generic[Key, Value]):
    key: Key
    value: Value


MapTree = Optional[MapTreeLeaf[Key, Value]]


@dataclass
class MapTreeNode(MapTreeLeaf[Key, Value]):
    left: MapTree[Key, Value]
    right: MapTree[Key, Value]

    height: int


def partition_aux(
    predicate: Callable[[Key, Value], bool],
    m: MapTree[Key, Value],
    acc: Tuple[MapTree[Key, Value], MapTree[Key, Value]],
) -> Tuple[MapTree[Key, Value], MapTree[Key, Value]]:
    if isinstance(m, MapTreeNode):
        ret = partition_aux(predicate, m.right, acc)
        return ret
    else:
        return acc
```

**VS Code extension or command-line**

Pylance: v2022.5.2, Pyright 1.1.247

**Comment by erictraut** [comment]
I agree this should work, but it may be difficult to fix. I'll need to give this additional thought. The challenge is in distinguishing between the type arguments in the outer function and the inner function since they have the same name and scope.

It's interesting to note that mypy evalutes a type that includes `Any` type arguments.
```
Revealed type of "ret" is "Tuple[Union[MapTreeLeaf[Any, Any], None], Union[MapTreeLeaf[Any, Any], None]]"
```
So pyright and mypy agree on the resulting type. However, I think both of them are arguably wrong.

**Comment by erictraut** [comment]
Here's a minimal sample that repros the problem:

```python
# pyright: strict
from typing import TypeVar
_T = TypeVar("_T")

def func(predicate: list[_T]) -> list[_T]:
    return func(predicate)
```

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.250, which I just published. It will also be included in the next release of pylance.
